File: Java/li1/plp/imperative1/Exemplo4.java
package li1.plp.imperative1;

import li1.plp.expressions2.expression.ExpEquals;
import li1.plp.expressions2.expression.Id;
import li1.plp.expressions2.expression.ValorInteiro;
import li1.plp.expressions2.expression.ValorString;
import li1.plp.imperative1.command.Atribuicao;
import li1.plp.imperative1.command.Comando;
import li1.plp.imperative1.command.ComandoDeclaracao;
import li1.plp.imperative1.command.IfThenElse;
import li1.plp.imperative1.command.SequenciaComando;
import li1.plp.imperative1.command.Write;
import li1.plp.imperative1.declaration.DeclaracaoComposta;
import li1.plp.imperative1.declaration.DeclaracaoVariavel;
import li1.plp.imperative1.memory.ContextoCompilacaoImperativa;
import li1.plp.imperative1.memory.ContextoExecucaoImperativa;
import li1.plp.imperative1.memory.ListaValor;

public class Exemplo4 {

    public static void main(String[] args) {
        Id idN = new Id("n");
        Id idM = new Id("m");

        // var n = 0, var m = 0
        DeclaracaoVariavel decN0 = new DeclaracaoVariavel(idN, new ValorInteiro(0));
        DeclaracaoVariavel decM0 = new DeclaracaoVariavel(idM, new ValorInteiro(0));
        DeclaracaoComposta decComposta = new DeclaracaoComposta(decN0, decM0);

        // n := 2
        Comando atribN2 = new Atribuicao(idN, new ValorInteiro(2));

        // m := 3
        Comando atribM3 = new Atribuicao(idM, new ValorInteiro(3));

        // if (m == n) then write(...) else write(...)
        ExpEquals condicaoIf = new ExpEquals(idM, idN);
        Comando thenBranch = new Write(new ValorString("valores de entrada iguais"));
        Comando elseBranch = new Write(new ValorString("valores de entrada diferentes"));
        Comando condicionalIf = new IfThenElse(condicaoIf, thenBranch, elseBranch);

        // Juntando os comandos em sequência: n := 2; m := 3; condicionalIf
        Comando seqComandos2 = new SequenciaComando(atribM3, condicionalIf);
        Comando sequenciaCompleta = new SequenciaComando(atribN2, seqComandos2);

        // Bloco do comando declaração completo
        Comando blocoDeclaracao = new ComandoDeclaracao(decComposta, sequenciaCompleta);

        Programa programa = new Programa(blocoDeclaracao);

        try {
            ContextoCompilacaoImperativa ambComp = new ContextoCompilacaoImperativa(new ListaValor());
            if (programa.checaTipo(ambComp)) {
                ContextoExecucaoImperativa ambExec = new ContextoExecucaoImperativa(new ListaValor());
                System.out.println("Saida do Exemplo 4: " + programa.executar(ambExec));
            } else {
                System.out.println("Erro de tipo no Exemplo 4!");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: Python/loo1-python/Exemplo4.py
from expression.exp_equals import ExpEquals
from expression.id import Id
from expression.valor_inteiro import ValorInteiro
from expression.valor_string import ValorString
from command.atribuicao import Atribuicao
from command.comando_declaracao import ComandoDeclaracao
from command.if_then_else import IfThenElse
from command.sequencia_comando import SequenciaComando
from command.write import Write
from declaration.declaracao_composta import DeclaracaoComposta
from declaration.declaracao_variavel import DeclaracaoVariavel
from memory.contexto_compilacao_imperativa import ContextoCompilacaoImperativa
from memory.contexto_execucao_imperativa import ContextoExecucaoImperativa
from memory.lista_valor import ListaValor
from Programa import Programa
from exception.identificador_ja_declarado_exception import IdentificadorJaDeclaradoException
from exception.identificador_nao_declarado_exception import IdentificadorNaoDeclaradoException
from exception.entrada_vazia_exception import EntradaVaziaException
from exception.erro_tipo_entrada_exception import ErroTipoEntradaException

def main():
    idN = Id("n")
    idM = Id("m")

    # var n = 0, var m = 0
    decN0 = DeclaracaoVariavel(idN, ValorInteiro(0))
    decM0 = DeclaracaoVariavel(idM, ValorInteiro(0))
    decComposta = DeclaracaoComposta(decN0, decM0)

    # n := 2
    atribN2 = Atribuicao(idN, ValorInteiro(2))

    # m := 3
    atribM3 = Atribuicao(idM, ValorInteiro(3))

    # if (m == n) then write(...) else write(...)
    condicaoIf = ExpEquals(idM, idN)
    thenBranch = Write(ValorString("valores de entrada iguais"))
    elseBranch = Write(ValorString("valores de entrada diferentes"))
    condicionalIf = IfThenElse(condicaoIf, thenBranch, elseBranch)

    # Juntando os comandos em sequência: n := 2; m := 3; condicionalIf
    seqComandos2 = SequenciaComando(atribM3, condicionalIf)
    sequenciaCompleta = SequenciaComando(atribN2, seqComandos2)

    # Bloco do comando declaração completo
    blocoDeclaracao = ComandoDeclaracao(decComposta, sequenciaCompleta)

    programa = Programa(blocoDeclaracao)

    try:
        ambComp = ContextoCompilacaoImperativa(ListaValor())
        if programa.checaTipo(ambComp):
            ambExec = ContextoExecucaoImperativa(ListaValor())
            print("Saida do Exemplo 4:", programa.executar(ambExec))
        else:
            print("Erro de tipo no Exemplo 4!")
    except IdentificadorJaDeclaradoException as e:
        print("Erro!! Identificador já declarado!")
        print("Detalhes:", e)
    except IdentificadorNaoDeclaradoException as e:
        print("Erro!! Identificador não declarada!")
        print("Detalhes:", e)
    except EntradaVaziaException as e:
        print("Erro! Nenhuma entrada foi fornecida!")
        print("Detalhes:", e)
    except ErroTipoEntradaException as e:
        print("Erro! O tipo de dado na entrada nao corresponde!")
        print("Detalhes:", e)

if __name__ == '__main__':
    main()
